[12.0] [FIX] website_sale_product_minimal_price: Bug with formulas

CarlosRoca13 opened the pull request:
Fixed a bug that compute allways the formula prices from the price. Now the price take care about the `base` field on pricelist item ids.
I add an improvement to take care about the categies.

cc @Tecnativa TT24954

@carlosdauden @chienandalu this is the solution of the bug mentioned

CarlosRoca13 committed:
[FIX] website_sale_product_minimal_price: Bug with formulas
Fixed a bug that compute allways the formula prices from the price. Now the price
take care about the based on.
Improved implementation to take care about the categories.

sergio-teruel reviewed (approved):
Check travis...
Tested on runbot

pedrobaeza committed:
[FIX+IMP] website_sale_product_minimal_price: Proper computation + use core computation

* Use core product price computation instead of computing prices by hand.
* Proper computation of the combination info for the cheapest variant.

pedrobaeza commented:
@CarlosRoca13 @sergio-teruel check the new algorithm for the grid view prices and the selection of the combination with lowest price. I haven't touched the scaled price table though.

CarlosRoca13 committed:
[IMP] website_sale_product_minimal_price: Controller core computation
Set on controller the proper computation to set the correct scaled price.

pedrobaeza commented:
Is this still WIP?

CarlosRoca13 commented:
No sorry, ready to be merged

pedrobaeza commented on website_sale_product_minimal_price/controllers/main.py:
```diff
@@ -23,55 +22,25 @@ def get_combination_info_pricelist_atributes(
+        pricelist_items = pricelist.item_ids.filtered(
             lambda i: i.product_id.id == product.id
+            or i.product_tmpl_id.id == product.product_tmpl_id.id
+            or i.categ_id.id == product.categ_id.id
```
On pricelist items, an upper category in the hierarchy includes all children categories, so this is not correct at all.

pedrobaeza commented on website_sale_product_minimal_price/controllers/main.py:
```diff
@@ -23,55 +22,25 @@ def get_combination_info_pricelist_atributes(
         pricelist = current_website.get_current_pricelist()
-        pricelist_variant_items = pricelist.item_ids.filtered(
+        pricelist_items = pricelist.item_ids.filtered(
             lambda i: i.product_id.id == product.id
```
No need to compare with `.id`. `i.product_id == product.product_id` is the same

pedrobaeza commented on website_sale_product_minimal_price/controllers/main.py:
```diff
@@ -23,55 +22,25 @@ def get_combination_info_pricelist_atributes(
-        pricelist_variant_items = pricelist.item_ids.filtered(
+        pricelist_items = pricelist.item_ids.filtered(
             lambda i: i.product_id.id == product.id
+            or i.product_tmpl_id.id == product.product_tmpl_id.id
```
I think the order is relevant here, as having a rule for template and for variant makes one of them invalid, and you are not taking this into account here.

CarlosRoca13 replied:
I take care about that on `_prepare_dictionary` method.

pedrobaeza replied:
I'm still not seeing any difference on this matter.

Imagine this pricelist:

- Template with min_qty = 1 - Price 10
- Template with min_qty = 30 - Price 5
- Variant with min_qty = 1 - Price 11
- Variant with min_qty = 5  - Price 10
- Variant with min_qty = 40  - Price 6

My guess is that this is going to paint

>=1 - 11 €
>=1 - 10 €
>=5 - 10 €
>=30 - 5 €
>=40 - 6 €

sergio-teruel reviewed (approved):
Ready for merge??

CarlosRoca13 commented:
I solved what @pedrobaeza said and I think that it is ready to be merged.

sergio-teruel commented on website_sale_product_minimal_price/models/product_template.py:
```diff
@@ -1,120 +1,81 @@
+        add_qty = 0
+        has_distinct_price = False
+        for product in self.product_variant_ids:
+            for qty in [1, math.inf]:
```
@CarlosRoca13 I think that the float_compare method crash with inf... @pedrobaeza 2 days ago changed inf to 99999...

sergio-teruel committed:
[IMP] website_sale_product_minimal_price: Compute price for variants only for variants defined in pricelist items.
TT27030

CarlosRoca13 committed:
[FIX] website_sale_product_minimal_price: Bug at the js inheritance
Solved the js inheritance deleting the contition that just allow to execute the parent code when the product hasn't items on the pricelist.

pedrobaeza commented on website_sale_product_minimal_price/models/product_template.py:
```diff
@@ -1,120 +1,96 @@
+            # remove returned values that are variant specific
+            res.filtered(lambda x: x.attribute_id not in variant_attributes)
+            # and inject cheapest variant ones
+            res += ptavs
```
This is not being sorted. See https://github.com/OCA/e-commerce/pull/460/files#diff-4df29720f5325f6a08d6648efb0e935de9b9cf38bc79e27087fb4c3e8f958c69L118

Why previous code was not correct (filtering if you want if pricelist is present).

chienandalu reviewed (commented):
@CarlosRoca13 what's to do here?

chienandalu commented:
What's left to do here. This has been in production for months isn't it?

pedrobaeza commented:
My comment is still unattended. The sort thing is incorrect, although it's difficult that somebody is aware of it.

CarlosRoca13 commented:
@pedrobaeza That changes that you say, were added by you, and I'm seeing that on the product sheet the order is correct. I found a problem on products grid that don't take the minimal value, but is not produced by that fragment of code. I'm working for solve it as soon as possible.

pedrobaeza commented:
OK, I see. In some point, my code was not used, but it's not the case now.

CarlosRoca13 committed:
[FIX] website_sale_product_minimal_price: Minimal price compute fixed

CarlosRoca13 commented:
@pedrobaeza I think that the problem that you mean is solved 🥳

pedrobaeza commented:
/ocabot merge patch

pedrobaeza commented:
/ocabot merge patch
